### ISSUE TITLE
Subscribing to tx event works for multiple kvpairs using same key

### ISSUE THREAD

**Issue by ethanfrey** [tendermint/tendermint#1385]
**Is this a BUG REPORT or FEATURE REQUEST?** (choose one):

Feature Request

**Tendermint version** (use `tendermint version` or `git rev-parse --verify HEAD` if installed from source):

Last tested on rpc bugfix branch
`0.17.0-34c800c1`
(but also with 0.16.1 and 0.17.1)

**ABCI app** (name for built-in, URL for self-written if it's publicly available):

https://github.com/confio/weave - mycoind

**Use Case**

I have a wallet and want to be notified of changes, so I can subscribe to all tx that affect the balance.

Ideally, I would do something like `tm.event = 'Tx' AND wallet='<my hex address>'`. However, this would require the transaction to create two events `wallet=<sender hex address>` and `wallet=<recipient hex address>`. This is currently not supported.

A previous release of cosmos-sdk:basecoind resolved this by using one tag for sender and one for recipient. This is a temporary work-around, but brings up two issues. First, because queries do not support OR, we need to make two separate subscriptions (`sender=<my hex address>` and `recipient=<my hex address>`). These both will make separate callbacks, and when I send some cash to myself it will show up twice. (This is a bit annoying, ask @faboweb and his desire for the txhash on subscriptions to detect duplicates, now fixed #1367).

A bigger issue is multi-account sends. `basecoind` is set up to allow many input accounts and many output accounts, similar to the bitcoin model, which allows for secure swaps in one tx with multiple signatures. However, if there ever was a transaction with two senders, the above work-around wouldn't work and it is impossible to subscribe to all multi-output transactions that affect your account.

Also, please note that the `/search_tx` functionality currently supports the same key showing up on a given tx multiple times with different values, and searching for any of them will match the tx. Making this change will maintain consistency between the search and subscribe interfaces.

**Proposed Changes**

Look at https://github.com/tendermint/tendermint/blob/develop/types/event_bus.go#L88-L120

`func (b *EventBus) PublishEventTx(event EventDataTx) error` builds up a `map[string]interface{}` of all the tags. The contents are currently either `string` or `int64` (only for height). (Actually also an unmatchable `[]byte` but that should be fixed in another issue).

I would propose another case `[]string`, which is used on all `event.Result.Tags`. When matching, the subscription will match if the query matches any of the strings in the slice. We still require multiple kvpairs and do not complicate that interface, we only aggragate them when building the map in `PublishEventTx`.

**Dependencies**

The fact that subscribing to delivertx tags works at all depends upon locally applying the patch that @melekes identified as part of https://github.com/tendermint/tendermint/issues/1369#issuecomment-376852991

I would be fine writing and submitting a PR once a proper solution to that bug is merged into develop, and there is agreement that this feature is desired.

**Comment by ethanfrey** [comment]
@melekes any decisions here or on #1369 ?

If there was a discussion as to desired solution, I could start a proposal for a fix that I could make as a PR. But I don't want to work if the PR will be rejected.

**Comment by ethanfrey** [comment]
Okay, I just consider this part of a larger refactor of the whole search/subscribe package, and don't wait for a small fix.

I will work around this behavior for now and wait for the improvements that eventually come. Just please link this issue and close it when `OR` `array.include?` or something similar is implemented.

**Comment by ethanfrey** [comment]
I think it would not be very hard to change https://github.com/tendermint/tendermint/blob/develop/types/event_bus.go#L93 to `map[string][]string` (nicer than the interface{} there last time I looked).

And just append to the array in https://github.com/tendermint/tendermint/blob/develop/types/event_bus.go#L102

Update the TagMap.Get() to return `[]string`: https://github.com/tendermint/tendermint/blob/develop/libs/pubsub/pubsub.go#L76

And then update the queries to handle the multiple choices: https://github.com/tendermint/tendermint/blob/develop/libs/pubsub/query/query.go

If you really want this feature, @ebuchman , I am happy to put some work on a PR, but it will probably need some help/polish from others.

**Comment by melekes** [comment]
Closing in favor of https://github.com/tendermint/tendermint/issues/1859.
